module/jobqueue.go

```diff
@@ -0,0 +1,29 @@
+package module
+
+const (
+	ConsumeProgressVerificationBlockHeight = "ConsumeProgressVerificationBlockHeight"
+	ConsumeProgressVerificationChunkIndex  = "ConsumeProgressVerificationChunkIndex"
+)
+
+// JobID is a unique ID of the job.
+type JobID string
+
+// JobConsumer consumes jobs from a job queue, and it remembers which job it has processed, and is able
+// to resume processing from the next.
+type JobConsumer interface {
+	// Start starts processing jobs from a job queue. If this is the first time, a processed index
+	// will be initialized in the storage. If it fails to initialize, an error will be returned
+	Start(defaultIndex int64) error
+
+	// Stop gracefully stops the consumer from reading new jobs from the job queue. It does not stop
+	// the existing worker finishing their jobs
+	Stop()
```

Comment:
I need this `Stop` method, because otherwise the existing worker will notify the job is done by the callback, and the callback will call `Check` to trigger new job, and then this ever will never end. The `Stop` will break this loop by updating the flag and not triggering any new job.

Reply:
How do we make sure that worker goroutines spawned by this component stop gracefully when we shut down the node?

Reply:
When we stop the node, we [will stop the consumer](https://github.com/onflow/flow-go/pull/361/files#diff-89b155078f45a7c7ae158ee1c57d7580d6b9160d96f02ba3818ccc3d16e5e38dR156), which enters the graceful shutdown period. During this, the worker will still keep working on the existing jobs. When the existing jobs are finished, the worker will call `Check`, since we've called `Stop` on the consumer, the consumer will not pass any job to the worker. This [test case](https://github.com/onflow/flow-go/pull/330/files#diff-59177c7a84790d9c5098c57ffff4dadf5e1d99d49fb04e553f356a0466433abcR368-R387) also verifies this behavior

Reply:
Oh, I see what you meant. You are right. When `Stop` is called, we actually [exit right away](https://github.com/onflow/flow-go/pull/361/files#diff-89b155078f45a7c7ae158ee1c57d7580d6b9160d96f02ba3818ccc3d16e5e38dR159-R160), without waiting for the worker to finish the existing jobs.

Reply:
Although, I think it's still not the `consumer`'s job to block the worker or return a channel or something for the caller to wait, because that would couple the consumer and the worker. 

We could let the caller stop the `consumer` and `worker` or `engine` separately, since the caller must know both the consumer and the worker.

In the case of `BlockConsumer`, since both `assigner.BlockConsumer` and `assigner.Engine` are ReadyDownAware component, when we shut down the node, we will call the `Done` method on both of them. The Done method on `BlockConsumer` will stop the consumer and ensure not more job will be passed to the engine, and the `Done` on the engine will wait until the existing jobs are done. Since we anyway have to wait for the engine to finish its own job.

Reply:
Note graceful shutdown is nice to have, but not required. Because the consumer won't update it's update the processed index in storage unless the worker has notified the consumer. 

So once we stop the consumer, it is still able to update the processed index when got notified, but even if the node crashes right away, it will still resume from the next unprocessed index, which might work on the same job twice, but at least guaranteed "at least one delivery" and won't miss any job.

Also the job queue doesn't allow duplication, so if a worker's job is to read from one job queue, and add a new job to another queue, then working on the same job twice, won't create duplicated jobs in the downstream job queue. 

For instance, processing the same finalized block from the block queue twice wouldn't save duplicated chunks in the chunks queue.

Reply:
> Although, I think it's still not the consumer's job to block the worker or return a channel or something for the caller to wait, because that would couple the consumer and the worker.

Since the consumer is the one spawning the goroutine, it should be the one responsible for stopping it. Otherwise we have cross-API contracts that certain functions in module 1 (the engine) will be invoked as goroutines by module 2 (the job queue) without that being explicit in the engine, but the engine is still responsible for stopping these goroutines. 

It's true that graceful shutdown is not required for the specific behaviour of the worker where this is currently used. But for a component that is intended to be generally applicable I feel we should invest in making it more resilient. 

I think this is fairly simple to implement with a `WaitGroup`:
* in `Start`, add one to the waitgroup
* add one to the wait group for each running goroutine
* in `Stop`, decrement one from the waitgroup, the call `wg.Wait`

Reply:
Good point. Added.